src/txrequest.h

```diff
@@ -0,0 +1,213 @@
+     *
+     * This should be called when a transaction is successfully added to the mempool, seen in a block, or for
+     * whatever reason we no longer care about it. The is_wtxid flag of gtxid is ignored.
+     */
+    void ForgetTx(const GenTxid& gtxid);
+
+    /** Adds a new CANDIDATE entry.
+     *
+     * Does nothing if one already exists for that (txhash, peer) combination (whether it's CANDIDATE, REQUESTED, or
+     * COMPLETED). Note that this means a second INV with the same txhash from the same peer will be ignored, even
+     * if one is a txid and the other is wtxid (but that shouldn't happen, as BIP339 requires that all announced
+     * inventory is exclusively using MSG_WTX). The new entry is given the specified preferred and reqtime values,
+     * and takes it is_wtxid from the specified gtxid. It is eligible to get a first marker if overloaded is false
+     * (but also subject to the other rules regarding the first marker).
+     */
+    void ReceivedInv(uint64_t peer, const GenTxid& gtxid, bool preferred, bool overloaded,
+        std::chrono::microseconds reqtime);
+
+    /** Converts the CANDIDATE entry for the provided peer and gtxid into a REQUESTED one.
+     *
+     * Expiry is set to the specified value. This can ONLY be called immediately after GetRequestable was called
+     * (for the same peer), with only ForgetTx and other RequestedTx calls (both for other txhashes) in
+     * between. Any other non-const operation removes the ability to call RequestedTx.
+     */
+    void RequestedTx(uint64_t peer, const GenTxid& gtxid, std::chrono::microseconds expiry);
+
+    /** Converts any CANDIDATE or REQUESTED entry to a COMPLETED one, if one exists.
+     *
+     * It should be called whenever a transaction or NOTFOUND was received from a peer. When a good transaction is
+     * received, ForgetTx should be called instead of (or in addition to) this operation.
+     */
+    void ReceivedResponse(uint64_t peer, const GenTxid& gtxid);
+
+    // The operations below inspect the data structure.
+
+    /** Find the txids to request now from peer.
+     *
+     * It does the following:
+     *  - Convert all REQUESTED entries (for all txhashes/peers) with (expiry <= now) to COMPLETED entries.
+     *  - Requestable entries are selected: CANDIDATE entries from the specified peer with (reqtime <= now) for
+     *    which the specified peer is the best choice among all such CANDIDATE entries with the same txhash (subject
+     *    to preference/first rules, and tiebreaking using a deterministic salted hash of peer and txhash).
+     *  - The selected entries are sorted in order of announcement (even if multiple were added at the same time, or
+     *    even when the clock went backwards while they were being added), converted to GenTxids using their
+     *    is_wtxid flag, and returned.
+     */
+    std::vector<GenTxid> GetRequestable(uint64_t peer, std::chrono::microseconds now);
```

Comment:
```suggestion
    std::vector<GenTxid> ExpireReqestedAndGetRequestable(uint64_t peer, std::chrono::microseconds now);
```

to make it less reliant on comments for intent

Reply:
I'm not entirely sure, as the caller really doesn't care about the fact that it also expires things. It just wants to know what should be requested.